### PR TITLE
Implement URL replacement for client ids available to amp-analytics

### LINKED ISSUE #1124: Add clientId support to amp-analytics
The clientId API (`src/cid.js`) requires parameters that specify the client id's namespace and consent. 

We need a way to express these parameters in amp-analytics configuration blocks, and need to determine when and how consent is presented.

### LINKED ISSUE #1141: amp-analytics should use layoutCallback instead of buildCallback
This automatically prevents it from running while doc is in prerender mode. We want to keep CPU usage in that time as low as possible which is why it should be delayed.

CC @rudygalfi

### PR DESCRIPTION
**Pull request by cramforce**
Also makes the following changes

- Switches `amp-analytics` to use layoutCallback instead of `buildCallback`. Closes #1141 and makes the tests more straight forward since they are now async.
- Support for promised based URL replacement
- Support for (right now only one) arguments to URL replacement functions.

Related to #961 #871
Fixes #1124

### CODE REVIEW

**Comment by cramforce** [comment]
CC @btownsend @avimehta

File: builtins/amp-pixel.js [hunk 1]
```diff
@@ -44,16 +44,17 @@ export function installPixel(win) {
...
+      const src = this.element.getAttribute('src');
+      return urlReplacementsFor(this.getWin()).expand(this.assertSource(src))
+          .then(src => {
+            const image = new Image();
+            image.src = src;
+            image.width = 1;
+            image.height = 1;
+            // Make it take zero space
+            this.element.style.width = 0;
+            this.element.appendChild(image);
```
Comment (dvoytenko) [inline builtins/amp-pixel.js]: Why are we even adding image to DOM?

File: extensions/amp-list/0.1/amp-list.js [hunk 2]
```diff
@@ -56,19 +56,20 @@ export class AmpList extends AMP.BaseElement {
...
-      const items = data['items'];
-      return templatesFor(this.getWin()).findAndRenderTemplateArray(
-          this.element, items).then(this.rendered_.bind(this));
-    });
+    return this.urlReplacements_.expand(assertHttpsUrl(
+        this.element.getAttribute('src'), this.element)).then(src => {
+          const opts = {
+            credentials: this.element.getAttribute('credentials')
+          };
+          return xhrFor(this.getWin()).fetchJson(src, opts).then(data => {
```
Comment (dvoytenko) [inline extensions/amp-list/0.1/amp-list.js]: Chaining, not nesting.

File: src/url-replacements.js [hunk 3]
```diff
@@ -82,7 +83,16 @@ class UrlReplacements {
     // single page view. It should have sufficient entropy to be unique for
     // all the page views a single user is making at a time.
     this.set_('PAGE_VIEW_ID', () => {
-      documentInfoFor(this.win_).pageViewId;
+      return documentInfoFor(this.win_).pageViewId;
+    });
+
+    this.set_('CLIENT_ID', (data, name) => {
+      return cidFor(this.win_).then(cid => {
```
Comment (dvoytenko) [inline src/url-replacements.js]: This looks like we introduce an implicit dependency between elements and js binaries. We have to be careful to ensure that any plans to "optimize" away unused `<script>` would not break things.

File: src/url-replacements.js [hunk 4]
```diff
@@ -111,15 +119,42 @@ class UrlReplacements {
    */
   expand(url, opt_data) {
```
Comment (dvoytenko) [inline src/url-replacements.js]: Expand's return type must change, right?

File: src/url-replacements.js [hunk 5]
```diff
@@ -111,15 +119,42 @@ class UrlReplacements {
...
       if (!val && val !== 0) {
         val = '';
       }
       return encodeURIComponent(val);
+    };
+    url = url.replace(expr, (match, name, arg) => {
+      const val = this.replacements_[name](opt_data, arg);
+      // In case the produced value is a promise, we don't actually
+      // replace anything here, but do it again when the promise resolves.
+      if (val instanceof Promise) {
```
Comment (dvoytenko) [inline src/url-replacements.js]: That could in theory run into the same problem we had with Promise polyfill and native like we had with fetch. I just recently cleaned those up. Should we err here as well?

File: src/url-replacements.js [hunk 6]
```diff
@@ -132,7 +167,8 @@ class UrlReplacements {
       for (const k in this.replacements_) {
         all += (all.length > 0 ? '|' : '') + k;
       }
-      this.replacementExpr_ = new RegExp('\\$?(' + all + ')', 'g');
+      this.replacementExpr_ =
+          new RegExp('\\$?(' + all + ')(?:\\(([a-zA-Z-_]+)\\))?', 'g');
```
Comment (dvoytenko) [inline src/url-replacements.js]: Why is this necessary?

File: src/url-replacements.js [hunk 7]
```diff
@@ -111,15 +119,42 @@ class UrlReplacements {
...
+    url = url.replace(expr, (match, name, arg) => {
+      const val = this.replacements_[name](opt_data, arg);
+      // In case the produced value is a promise, we don't actually
+      // replace anything here, but do it again when the promise resolves.
+      if (val instanceof Promise) {
+        const p = val.then(v => {
+          url = url.replace(match, encodeValue(v));
+        });
+        if (replacementPromise) {
+          replacementPromise = replacementPromise.then(() => {
```
Comment (dvoytenko) [inline src/url-replacements.js]: `replacementPromise.then(() => p)`

File: src/url-replacements.js [hunk 8]
```diff
@@ -111,15 +119,42 @@ class UrlReplacements {
...
+        } else {
+          replacementPromise = p;
+        }
+        return match;
+      }
+      return encodeValue(val);
     });
+
+    if (replacementPromise) {
+      replacementPromise = replacementPromise.then(() => {
```
Comment (dvoytenko) [inline src/url-replacements.js]: ditto on shorthand

File: src/url-replacements.js [hunk 9]
```diff
@@ -111,15 +119,42 @@ class UrlReplacements {
...
       }
       return encodeURIComponent(val);
+    };
+    url = url.replace(expr, (match, name, arg) => {
+      const val = this.replacements_[name](opt_data, arg);
+      // In case the produced value is a promise, we don't actually
+      // replace anything here, but do it again when the promise resolves.
+      if (val instanceof Promise) {
+        const p = val.then(v => {
+          url = url.replace(match, encodeValue(v));
```
Comment (dvoytenko) [inline src/url-replacements.js]: This has a slight negative that two RANDOM would produce the same value when replaced. But this is somewhat of a contrived example. Alternatively, we could insert some autogenerated match string (REPLACE_ME_1,2,3) and promises when return would put values into this map and after all promises returned, it would do one set of replacements from this map. But, if it doesn't seem like an issue, we can just keep as is.

File: extensions/amp-list/0.1/amp-list.js [hunk 10]
```diff
@@ -56,19 +56,20 @@ export class AmpList extends AMP.BaseElement {
...
-    };
-    return xhrFor(this.getWin()).fetchJson(src, opts).then(data => {
-      assert(typeof data == 'object' && Array.isArray(data['items']),
-          'Response must be {items: []} object %s %s',
-          this.element, data);
-      const items = data['items'];
-      return templatesFor(this.getWin()).findAndRenderTemplateArray(
-          this.element, items).then(this.rendered_.bind(this));
-    });
+    return this.urlReplacements_.expand(assertHttpsUrl(
```
Comment (dvoytenko) [inline extensions/amp-list/0.1/amp-list.js]: How does this work with amp-list? Is amp-list allowed to use CID in its URLs? Does amp-list now become implicitly dependent on amp-analytics?

File: builtins/amp-pixel.js (hunk 1 above)
Comment (cramforce) [inline builtins/amp-pixel.js]: If any reason, just for testing. No good reason to change in this CL.

File: extensions/amp-list/0.1/amp-list.js (hunk 10 above)
Comment (cramforce) [inline extensions/amp-list/0.1/amp-list.js]: It does. But only if you expand CLIENT_ID. The error message needs to be good, of course.

I'll look into improving it a bit more.

File: extensions/amp-list/0.1/amp-list.js (hunk 2 above)
Comment (cramforce) [inline extensions/amp-list/0.1/amp-list.js]: Sorry, don't understand what you mean.

File: src/url-replacements.js (hunk 3 above)
Comment (cramforce) [inline src/url-replacements.js]: Yeah, I've thought about that as well. We definitely don't want CID in core, though.

File: extensions/amp-list/0.1/amp-list.js (hunk 10 above)
Comment (dvoytenko) [inline extensions/amp-list/0.1/amp-list.js]: Ok. Good. If we can provide a good warning or error - it should be fine.

File: src/url-replacements.js (hunk 4 above)
Comment (cramforce) [inline src/url-replacements.js]: Done

File: src/url-replacements.js (hunk 5 above)
Comment (cramforce) [inline src/url-replacements.js]: Switched to thenable detection. Better?

File: extensions/amp-list/0.1/amp-list.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-list/0.1/amp-list.js]:
```
return urlReplacements.expand(url)
  .then(DO_XHR)
  .then(PROCESS_XHR)
```
Instead of 
```
return urlReplacements.expand(url)
  .then(DO_XHR.then(PROCESS_XHR))
```

File: src/url-replacements.js (hunk 9 above)
Comment (cramforce) [inline src/url-replacements.js]: Lets keep. Also, not sure actually. The replace only replaces one instance.

File: src/url-replacements.js (hunk 7 above)
Comment (cramforce) [inline src/url-replacements.js]: Nice

File: src/url-replacements.js (hunk 8 above)
Comment (cramforce) [inline src/url-replacements.js]: Done

File: src/url-replacements.js (hunk 3 above)
Comment (dvoytenko) [inline src/url-replacements.js]: Did you think about adding `<script>` tag ourselves if it's not present?

File: src/url-replacements.js (hunk 5 above)
Comment (dvoytenko) [inline src/url-replacements.js]: I think it is.

File: src/url-replacements.js (hunk 6 above)
Comment (cramforce) [inline src/url-replacements.js]: Added comment

File: src/url-replacements.js (hunk 9 above)
Comment (dvoytenko) [inline src/url-replacements.js]: Yes. That's true. I guess then nvm.

**Comment by cramforce** [comment]
PTAL

File: extensions/amp-list/0.1/amp-list.js (hunk 2 above)
Comment (dvoytenko) [inline extensions/amp-list/0.1/amp-list.js]: Did you see this one?

**Comment by dvoytenko** [comment]
LGTM

File: extensions/amp-list/0.1/amp-list.js (hunk 2 above)
Comment (cramforce) [inline extensions/amp-list/0.1/amp-list.js]: Done

File: src/url-replacements.js (hunk 3 above)
Comment (cramforce) [inline src/url-replacements.js]: That would be slow. We can still think about it in the future.

Comment (dvoytenko) [inline src/url-replacements.js]: Yes. LGTM.
